### ISSUE TITLE
Generate implementation traits to simplify use of `implement` macro

### ISSUE THREAD

**Issue by nerditation** [microsoft/windows-rs#1265]
short question: why are some methods of a interface are feature gated, while the interface itself is not, or gated with different feature set? I find it counter intuitive, because you can't author a COM class without the full definition of the vtable entries.

so the long story is, I was trying to write a class which implements ``Windows::Win32::UI::TextServices::ITextStoreACP`` interface, but the ``#[implement(...)]`` attribute generates many errors, some of which are

```console
error[E0412]: cannot find type `TS_ATTRVAL` in module `Windows::Win32::UI::TextServices`
  --> src\main.rs:22:1
   |
22 | / #[implement(
23 | |     Windows::Win32::UI::TextServices::ITfContextOwnerCompositionSink
24 | |     ,Windows::Win32::UI::TextServices::ITextStoreACP
25 | | )]
   | |__^ not found in `Windows::Win32::UI::TextServices`
   |
   = note: this error originates in the attribute macro `implement` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0412]: cannot find type `TS_ATTRVAL` in this scope
   --> src\main.rs:243:20
    |
243 |         paattrvals: *mut TS_ATTRVAL,
    |                          ^^^^^^^^^^ not found in this scope
```

I was very confused by these errors. because the [document](https://microsoft.github.io/windows-docs-rs/doc/bindings/Windows/Win32/UI/TextServices/struct.TS_ATTRVAL.html) says the type ``TS_ATTRVAL`` does exist in namespace ``Windows::Win32::UI::TextServices``

since the first one is from the expanded macro, and I don't know how the macro works, I have zero clue what was the cause; but the second one originated from the method signature inside my impl block, which I adapted from the binding [documents](https://microsoft.github.io/windows-docs-rs/doc/bindings/Windows/Win32/UI/TextServices/struct.ITextStoreACP.html#method.RetrieveRequestedAttrs):

```rust
fn RetrieveRequestedAttrs(
    &self,
    ulcount: u32,
    paattrvals: *mut TS_ATTRVAL,
    pcfetched: *mut u32
) -> Result<()> {
    todo!()
}
```

so I checked the source code, and found out the type ``TS_ATTRVAL``, and also the method ``RetrieveRequestedAttrs`` are behind the feature set ``#[cfg(all(feature="Win32_Foundation", feature="Win32_System_OleAutomation", feature="Win32_System_SystemServices"))]``, while the mod itself (and the interface ``ITextStoreACP``) is only gated by ``#[cfg(feature="Win32_UI_TextServices")]``.

after adding these features into ``Cargo.toml``, I am able to (at least) get the code compiled. but I feel this is really cumbersome.

to me, for instance, if I want to use the mod ``Win32::UI::TextServices``, I would like to just specify the ``"Win32_UI_TextServices"`` feature in dependencies section of ``Cargo.toml`` and be able to code. I'm fine with the "Win32_Foundation" one as it kind of like a prelude, but other than that, I think selecting features at the namespace/module granularity feels acceptable, but selecting features for each individual type, interface, even method? definitely not the correct granularity.

-----

oh, another issue writing this program, I would like to complaints a little bit about the ergonomics of the ``implement`` macro. it is very inconvenient and tedious to implement a interface, especially interfaces with many methods, because currently you don't get the usual assist from the tooling to do the scaffolding and boilerplates. you have to manually copy the methods' signatures from the document or source code, and maybe manually adjust the argument types, like reference, mut reference, ``Option``, etc.

to me, procedural macros are indistinguishable from black magic, I can't really give any technical suggestions, but still I'd like to imagine, if a COM interface were associated (somehow) with a Rust trait (what I call a guide trait) -- this trait doesn't neccessarily need to deal with ABI stuff like vtable, ffi marshalling, calling convention etc,  in fact we might never use it for dynamic dispatch -- since we have a Rust traits now, we can leverage all the tooling in the Rust ecosystem, e.g., use the editor to generate an ``impl`` block with all the methods' signatures boilerplates.

again, since I don't really know procedural macros in Rust, the following might be totally nonsense.

one possible solution is to generate the traits along side the COM interface abi types themselves, then the ``implement`` macro could generate generic stubs (think the ``XYZ_box`` stuff) against the traits instead of the concrete classes/structs.

this could potentially get rid of the restrictions on the naming (and wildcard ``use`` directive) of the rust mods in the current ``implement`` macro, since the ``XYZ_box`` stuff is generic over the traits instead of concrete type, they might reside in the same mod along with the interfaces and guide traits, which means the ``implement`` macro could accept actual valid Rust qualified identifiers instead of syntactical tokens.

**Comment by kennykerr** [comment]
Good feedback. 

The `cfg` features are tricky because making it more "automatic" means including many more modules automatically and kills build time in many cases. Making it more granular means having hundreds of thousands of features which is impractical. So there's a fine balance here that I'm still exploring.

The `implement` macro still needs a lot of work and I am considering requiring traits to make it easier to implement, but again that means generating a lot more code so I have to balance the effect on build time. 

This should get a lot easier to use over the coming months as I work through these issues.

**Comment by kennykerr** [comment]
Regarding the docs, they've now been updated to indicate which features are required to enable specific APIs:

https://microsoft.github.io/windows-docs-rs/doc/windows/Win32/UI/TextServices/struct.TS_ATTRVAL.html

**Comment by kennykerr** [comment]
Traits are coming: https://github.com/microsoft/windows-rs/issues/1345#issuecomment-1006971935
